emmarin/cl/ledger/src/ledger.rs

```diff
@@ -1,64 +1,66 @@
+
+use crate::bundle::ProvedBundle;
+use cl::zone_layer::{ledger::LedgerState, notes::ZoneId};
 
 #[derive(Debug, Clone)]
 pub struct ProvedLedgerTransition {
-    pub public: LedgerProofPublic,
     pub risc0_receipt: risc0_zkvm::Receipt,
 }
 
-// TODO: find a better name
-#[derive(Debug, Clone)]
-pub struct ProvedBundle {
-    pub balance: ProvedBalance,
-    pub ptxs: Vec<ProvedPartialTx>,
-}
-
-impl ProvedBundle {
-    fn to_public(&self) -> Vec<PtxPublic> {
-        self.ptxs.iter().map(|p| p.public.clone()).collect()
-    }
-
-    fn proofs(&self) -> Vec<risc0_zkvm::Receipt> {
-        let mut proofs = vec![self.balance.risc0_receipt.clone()];
-        proofs.extend(self.ptxs.iter().map(|p| p.risc0_receipt.clone()));
-        proofs
-    }
-}
-
 impl ProvedLedgerTransition {
-    pub fn prove(
-        ledger: LedgerWitness,
-        zone_id: ZoneId,
-        bundles: Vec<ProvedBundle>,
-        constraints: Vec<ConstraintProof>,
-    ) -> Result<Self> {
-        let witness = LedgerProofPrivate {
-            bundles: bundles.iter().map(|p| p.to_public()).collect(),
-            ledger,
+    pub fn prove(mut ledger: LedgerState, zone_id: ZoneId, bundles: Vec<ProvedBundle>) -> Self {
+        let mut witness = LedgerProofPrivate {
+            bundles: Vec::new(),
+            ledger: ledger.to_witness(),
             id: zone_id,
         };
 
         let mut env = risc0_zkvm::ExecutorEnv::builder();
 
-        for bundle in bundles {
-            for proof in bundle.proofs() {
-                env.add_assumption(proof);
+        // prepare the sparse merkle tree nullifier proofs
+        for proved_bundle in &bundles {
+            env.add_assumption(proved_bundle.risc0_receipt.clone());
+
+            let bundle = proved_bundle.public();
+            println!("OUTSIDE LEDGER_PROOF {:?}", bundle);
+            println!("BUNDLE_ID {:?}", nomos_cl_bundle_risc0_proof::BUNDLE_ID);
+
+            let zone_ledger_update = bundle
+                .zone_ledger_updates
+                .get(&zone_id)
+                .expect("why are we proving this bundle for this zone if it's not involved?");
+
+            let cm_root_proofs =
+                BTreeMap::from_iter(zone_ledger_update.cm_roots.iter().map(|root| {
+                    // We make the simplifying assumption that bundle proofs
+                    // are done w.r.t. the latest MMR (hence, empty merkle proofs)
+                    //
+                    // We can remove this assumption by tracking old MMR roots in the LedgerState
+                    (*root, vec![])
+                }));
+
+            let mut nf_proofs = Vec::new();
+            for nf in &zone_ledger_update.nullifiers {
+                let nf_proof = ledger.add_nullifier(*nf);
+                nf_proofs.push(nf_proof);
             }
+
+            for cm in &zone_ledger_update.commitments {
+                ledger.add_commitment(cm);
```

Comment:
yeah, true. It's because we have the limitation that we assume the roots are part of latest MMR, we don't use this, otherwise we would need to update the MMR for each bundle and then generate the root extension to the current MMR roots.

Reply:
I've removed this loop for now, but we'll have to address this when we support in the ledger state old roots.